### ISSUE TITLE
Level 3: Is there any other way to implement isTaken()?

### ISSUE THREAD

**Issue by bryankwe** [nus-cs2030/2122-s2#239]
Hi, from the previous threads on Level 3, many of you guys suggested having an `isTaken()` method to check if a Sword has been taken.

However, a Room contains an `ImList<Thing>`, and hence any `Sword` will be a `Thing` (which means we can't call `Sword` specific methods from the contents of this list)

Hence, my isTaken() method was implemented in my interface `Thing` (and now I call it `interacted()` instead). 

This works, but I was wondering if there was a better way to implement `isTaken()` since it feels redundant to have all classes implementing `Thing` to have the `interacted()` method (For instance, a Candle will not have any interaction)

I saw that some people explicitly mentioned that their `isTaken()` method is in the `Sword` class. I would really like to know how that was implemented. Thanks!

**TLDR**
Are there any other ways to implement isTaken() other than making it a method of your Thing interface/abstract class?

Thanks in advance to anyone who shares their views!

**Comment by lukasleeyo** [comment]
Hi there, for me I only implemented ``isTaken()`` inside my sword class. This is how I do it.

1. Implement ``isTaken()`` inside Sword class only.
2. When you loop through ``ImList<Thing>``, you check if that thing is a Sword or not using ``identify()`` method.
3. If it's a sword, do an explicit type casting e.g. ``((Sword) thing).isTaken()``.

You may be wondering that why I did an explicit type casting and maybe you thought is not allowed. But this is how I see it. Because if you use ``identify()`` to check if it's a sword, then it must be a Sword object since only Sword object returns the word "Sword" in ``identify()``. And we do explicit type casting so that we can access the ``isTaken()`` method inside sword class.

By doing all these, you avoid writing ``isTaken()`` as an abstract/interface methods, which makes sense because it would be weird if you have ``isTaken()`` sword status implemented in Thing since it only applies to Sword. 

Hope it helps! :)

**Comment by szeying02** [comment]
Oh I used a very strange method, that may not have adhered strictly to the tell-don't-ask principle. 

I created a boolean variable within the Room itself. As the game can only assumes that there are no duplicated items in the room. So the boolean is passed when creating new Rooms.

**Comment by StanleyMak** [comment]
Im not sure if my method is correct but what i did was to have a boolean attribute in my Sword class, 
private final boolean isTaken;

and i have a take() method that returns a new Sword with the boolean set to true;
similarly for my drop() method, it sets the boolean to false.

**Comment by bryankwe** [comment]
> Im not sure if my method is correct but what i did was to have a boolean attribute in my Sword class, private final boolean isTaken;
> 
> and i have a take() method that returns a new Sword with the boolean set to true; similarly for my drop() method, it sets the boolean to false.

@StanleyMak Thanks for sharing. But how did u manage to call isTaken() (a Sword method) on a Thing object? Is it similar to how @lukasleeyo did it? 

Also, thanks @lukasleeyo @szeying02 for sharing! Very thought-invoking :)

**Comment by jonyeokj** [comment]
> Hi there, for me I only implemented `isTaken()` inside my sword class. This is how I do it.
> 
> 1. Implement `isTaken()` inside Sword class only.
> 2. When you loop through `ImList<Thing>`, you check if that thing is a Sword or not using `identify()` method.
> 3. If it's a sword, do an explicit type casting e.g. `((Sword) thing).isTaken()`.
> 
> You may be wondering that why I did an explicit type casting and maybe you thought is not allowed. But this is how I see it. Because if you use `identify()` to check if it's a sword, then it must be a Sword object since only Sword object returns the word "Sword" in `identify()`. And we do explicit type casting so that we can access the `isTaken()` method inside sword class.
> 
> By doing all these, you avoid writing `isTaken()` as an abstract/interface methods, which makes sense because it would be weird if you have `isTaken()` sword status implemented in Thing since it only applies to Sword.
> 
> Hope it helps! :)

I did this at first, but in the end, I had to resort to defining isTaken() in my interface/abstract class because I had to keep track of which items were taken or not, and it didn't make sense to typecast my items as Swords to check.

Initially, I checked if the thing was a sword when picking it up, using Sword sword = (Sword) thing - but I realized if I wanted to introduce other things that could be picked up and dropped, I would have to check for those classes as well. Do you have an implementation whereby introducing other things that can be picked up/dropped works?

**Comment by alwinngjw** [comment]
Hi everyone, i had a similar implementation with @lukasleeyo  where I only have the isTaken() method inside my Sword class.

This is because i too find it weird to have a isTaken() method written in other items when it only applies to a Sword item
